### DISCUSSION TITLE
What is the correct way to implement the web version of <Protected> ?

### DISCUSSION THREAD

**Discussion by Albert-Gao** [expo/router#20] [Q&A]
trying to add a <Protected> layout wrapper for differentiating the authed and unauthed routes. It is all about web.

![image](https://user-images.githubusercontent.com/18282328/192891341-247be97d-4fb6-4d87-9a01-3777e6a8ed13.png)

![image](https://user-images.githubusercontent.com/18282328/192891426-4e9352d9-d4bf-4e1a-94bd-a8c79a2cc5be.png)

![image](https://user-images.githubusercontent.com/18282328/192891453-61823e87-c038-4bdf-986a-fae2be734aac.png)

## What is the expected behavior:
- on the login page, after a successful login, lead the user to `/dashboard`, which is protected
- access /dashboard directly from the URL, it should take the user back, or to the login page if can not go back

## What is the current behavior:
- on the login page, after a successful login, it CAN take the user to `/dashboard`
- access `/dashboard` directly from the URL with no auth state, it takes the user to `/`, and stuck at the element returned from <Protected> rather than honoring the `link.push('/login')`

This is what happened after accessing `/dashboard` from URL with non-auth state:

![image](https://user-images.githubusercontent.com/18282328/192891967-4e12a357-2bf6-48ff-9a72-ab3258100d0d.png)

wondering what i am missing here :)

related to this
https://github.com/expo/router/issues/14

**Comment by Albert-Gao** [comment]
Usually on the web, I would be something like this:

```javascript
function Protected({ children }) {
  const { isAuth } = useAuth()

  return isAuth ? children : <Navigate to={'/login'} />
}
```

The API for the <Naivgate> is here : https://reactrouter.com/en/main/components/navigate

It is just a componentized way to do the navigate, as you can see in the code above, i am trying to do something similar, if we have something like this, then my current folder structure would just work without any problem. Actually, it works, just the accessing directly  does not work, if we can make this part work, then we are all good here, you can wrap all authed routes with this layout, then you are done. :)

**Reply by EvanBacon** [reply]
I've added support for `link.replace()` in the `const link = useLink()` so you should be able to redirect without pushing anything on the stack.

Should be possible to implement `Navigate` like:

```tsx
import { useEffect } from 'react';
import { useLink } from 'expo-router';

export function Navigate({href}) {
  const link = useLink();
  
  useEffect(() => {
    link.replace(href);
  }, [href, link]);

  return null;
}

```

**Reply by Albert-Gao** [reply]
thanks, @EvanBacon 

tried it, but still not work, wonder what i am missing here:

to reproduce, `npx expo start --clear`
then try accessing `/dashboard` without login
it does not redirect back to `/login`

repo here:
https://github.com/Albert-Gao/expo-router-web-try-out/tree/web-auth-flow
